File: src/split/load-order.test.ts
import assert from "node:assert";
import { describe, it } from "node:test";
import { parseSync } from "@babel/core";
import type * as t from "@babel/types";
import {
  analyzeLoadOrder,
  bundleLoadOrderFacts,
  type LoadOrderFacts,
  orderRespectingLoadOrder
} from "./load-order.js";

/** Parse a module body into top-level statements. */
function stmts(src: string): t.Statement[] {
  const ast = parseSync(src, { sourceType: "unambiguous" });
  if (!ast || ast.type !== "File") throw new Error("parse failed");
  return ast.program.body;
}

function facts(src: string, pure?: string[]): LoadOrderFacts[] {
  return analyzeLoadOrder(stmts(src), {
    pureCallNames: pure ? new Set(pure) : undefined
  });
}

describe("analyzeLoadOrder — what a statement does while the module loads", () => {
  it("treats a function declaration as hoisted: no reads, no writes, no effects", () => {
    const [f] = facts(
      "function render(x) { return helper(x) + sideEffect(); }"
    );
    assert.strictEqual(f.hoisted, true);
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.reads], []);
    assert.deepStrictEqual([...f.writes], []);
  });

  it("records a literal declaration as a pure write", () => {
    const [f] = facts('var TAG = "[enforce]";');
    assert.strictEqual(f.hoisted, false);
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.writes], ["TAG"]);
    assert.deepStrictEqual([...f.reads], []);
  });

  it("does NOT read what a function/arrow BODY references — that runs later", () => {
    const [f] = facts("var run = () => { return moduleState + other(); };");
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.writes], ["run"]);
    assert.deepStrictEqual([...f.reads], []);
  });

  it("reads what an initializer EXPRESSION references — that runs now", () => {
    const [f] = facts("var derived = base.field;");
    assert.deepStrictEqual([...f.writes], ["derived"]);
    assert.deepStrictEqual([...f.reads], ["base"]);
  });

  it("flags an unknown call as an observable load-time effect", () => {
    const [f] = facts("var conn = openSocket();");
    assert.strictEqual(f.effects, true);
  });

  it("flags a VERIFIED pure wrapper call as effect-free (the lazy-init unlock)", () => {
    // `lazyInitializer` captures the generator in a closure and returns a thunk;
    // nothing observable happens until the thunk is invoked. The caller verifies
    // that shape structurally and passes the name in.
    const [f] = facts(
      "var cfg = lazyInitializer(() => buildConfig(globalSettings));",
      ["lazyInitializer"]
    );
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.writes], ["cfg"]);
    // the helper itself is read; the arrow body is not
    assert.deepStrictEqual([...f.reads], ["lazyInitializer"]);
  });

  it("recognises the pure wrapper through the emitted `(0, ns.helper)(...)` form", () => {
    const [f] = facts(
      "var cfg = (0, resourceLifecycle.lazyInitializer)(() => build());",
      ["lazyInitializer"]
    );
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.reads], ["resourceLifecycle"]);
  });

  it("keeps a non-pure call through the same form effectful", () => {
    const [f] = facts(
      "var cfg = (0, resourceLifecycle.runNow)(() => build());",
      ["lazyInitializer"]
    );
    assert.strictEqual(f.effects, true);
  });

  it("charges a bare `var` nothing — hoisted binding, no runtime effect", () => {
    const [f] = facts("var pending, queued;");
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.writes], []);
  });

  it("charges a bare `let` a write — TDZ makes its position observable", () => {
    const [f] = facts("let pending;");
    assert.deepStrictEqual([...f.writes], ["pending"]);
  });

  it("treats the module-exports registration call as an effect that reads its target", () => {
    const [f] = facts("defineModuleExports(memoryModule, { get: () => x });");
    assert.strictEqual(f.effects, true);
    assert.ok([...f.reads].includes("memoryModule"));
  });

  it("counts a top-level assignment to a module binding as a write, not an effect", () => {
    const [f] = facts("cachedValue = 1;");
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.writes], ["cachedValue"]);
  });

  it("counts an assignment THROUGH a member as an effect", () => {
    const [f] = facts("target.field = 1;");
    assert.strictEqual(f.effects, true);
  });

  it("treats a spread as an effect — it runs iterators and getters", () => {
    const [f] = facts("var merged = { ...source };");
    assert.strictEqual(f.effects, true);
  });

  it("treats a plain class declaration as an effect-free write that reads its superclass", () => {
    const [f] = facts(
      "class Panel extends Base { render() { return sideFx(); } }"
    );
    assert.strictEqual(f.effects, false);
    assert.deepStrictEqual([...f.writes], ["Panel"]);
    assert.deepStrictEqual([...f.reads], ["Base"]);
  });

  it("treats a class with a static initializer as effect-bearing — it runs at definition", () => {
    const [f] = facts("class Registry { static all = collectAll(); }");
    assert.strictEqual(f.effects, true);
  });

  it("treats control flow at the top level as effect-bearing", () => {
    assert.strictEqual(facts("if (flag) { boot(); }")[0].effects, true);
    assert.strictEqual(facts("try { boot(); } catch {}")[0].effects, true);
    assert.strictEqual(
      facts("for (var i = 0; i < 3; i++) tick();")[0].effects,
      true
    );
  });
});

describe("bundleLoadOrderFacts — the wrapper is verified by SHAPE, not by name", () => {
  // Bun's lazy-init helper, minified names and all. What makes it pure is the
  // `x && (y = x(x = 0))` body, not what it is called.
  const HELPER =
    "var qz = (gen, cached) => () => (gen && (cached = gen(gen = 0)), cached);";

  it("admits a call to the structurally verified helper as effect-free", () => {
    const src = [HELPER, "var mod = qz(() => loadHeavyThing());"].join("\n");
    const f = bundleLoadOrderFacts(stmts(src), src);
    assert.strictEqual(f[1].effects, false);
    assert.deepStrictEqual([...f[1].writes], ["mod"]);
  });

  it("keeps the same call effectful when the bundle has no such helper", () => {
    // Same call site, but nothing in this source has the lazy-init shape — so
    // `qz` is just an unknown function and the call must stay pinned.
    const src = [
      "var qz = makeThing;",
      "var mod = qz(() => loadHeavyThing());"
    ].join("\n");
    const f = bundleLoadOrderFacts(stmts(src), src);
    assert.strictEqual(f[1].effects, true);
  });
});

/** Deterministic shuffles — a property test that cannot flake. */
function seededShuffler(seed: number): (xs: number[]) => number[] {
  let state = seed;
  const rand = () => {
    state = (state * 1103515245 + 12345) % 2147483648;
    return state / 2147483648;
  };
  return (xs) => {
    const out = [...xs];
    for (let i = out.length - 1; i > 0; i--) {
      const j = Math.floor(rand() * (i + 1));
      [out[i], out[j]] = [out[j], out[i]];
    }
    return out;
  };
}

/** No non-hoisted statement may end up on the other side of an effect barrier. */
function assertNoBarrierCrossed(out: number[], f: LoadOrderFacts[]): void {
  const pos = new Map(out.map((s, i) => [s, i]));
  for (const b of out) {
    if (!f[b].effects) continue;
    for (const s of out) {
      if (s === b || f[s].hoisted) continue;
      assert.strictEqual(
        s < b,
        (pos.get(s) as number) < (pos.get(b) as number),
        `statement ${s} crossed barrier ${b}`
      );
    }
  }
}

/** A load-time read never floats above the statement that writes it. */
function assertNoReaderAboveWriter(out: number[], f: LoadOrderFacts[]): void {
  const pos = new Map(out.map((s, i) => [s, i]));
  for (const w of out) {
    for (const r of out) {
      if (w >= r || f[w].hoisted || f[r].hoisted) continue;
      if (!f[w].writes.some((n) => f[r].reads.includes(n))) continue;
      assert.ok(
        (pos.get(w) as number) < (pos.get(r) as number),
        `reader ${r} floated above writer ${w}`
      );
    }
  }
}

describe("orderRespectingLoadOrder — the legal permutations", () => {
  /** Apply a desired order and return the resulting statement order. */
  function order(src: string, desired: number[], pure?: string[]): number[] {
    const f = facts(src, pure);
    const slots = f.map((_, i) => i);
    return orderRespectingLoadOrder(slots, desired, f);
  }

  const PURE_FILE = [
    'var a = "one";',
    "var b = 2;",
    "var c = () => {};",
    "var d = /re/;"
  ].join("\n");

  it("permutes a file of pure declarations FREELY — every desired order is reached", () => {
    for (const desired of [
      [3, 2, 1, 0],
      [1, 3, 0, 2],
      [2, 0, 3, 1]
    ]) {
      assert.deepStrictEqual(order(PURE_FILE, desired), desired);
    }
  });

  it("never lets a reader precede the writer it depends on", () => {
    // `total` is written by stmt 0 and read at load time by stmt 1.
    const src = ["var total = 5;", "var doubled = total * 2;"].join("\n");
    assert.deepStrictEqual(order(src, [1, 0]), [0, 1]);
  });

  it("never lets a load-time read of a later write float above it (anti-dependency)", () => {
    const src = ["var view = state;", "var state = compute;"].join("\n");
    assert.deepStrictEqual(order(src, [1, 0]), [0, 1]);
  });

  it("keeps two writes to the same binding in order", () => {
    const src = ["counter = 1;", "counter = 2;"].join("\n");
    assert.deepStrictEqual(order(src, [1, 0]), [0, 1]);
  });

  it("never reorders effect-bearing statements relative to each other", () => {
    const src = ["boot();", "shutdown();"].join("\n");
    assert.deepStrictEqual(order(src, [1, 0]), [0, 1]);
  });

  it("never lets any statement CROSS an effect-bearing statement (the boot-crash rule)", () => {
    // This is the exact shape that crashed the runnable tree in exp037: the
    // registration call reads `mod`, which the declaration assigns at load time.
    const src = [
      "var mod = {};",
      "defineModuleExports(mod, { get: () => v });",
      'var tail = "t";'
    ].join("\n");
    assert.deepStrictEqual(order(src, [2, 1, 0]), [0, 1, 2]);
    assert.deepStrictEqual(order(src, [1, 0, 2]), [0, 1, 2]);
  });

  it("lets a HOISTED function cross anything, including an effect barrier", () => {
    const src = [
      "var mod = {};",
      "defineModuleExports(mod, {});",
      "function tail() {}"
    ].join("\n");
    // The function is wanted first; it is hoisted, so it may lead.
    assert.deepStrictEqual(order(src, [2, 0, 1]), [2, 0, 1]);
  });

  it("moves a pure lazy-init declaration across other pure declarations", () => {
    const src = [
      "var first = lazyInitializer(() => a());",
      "var second = lazyInitializer(() => b());",
      "var third = lazyInitializer(() => c());"
    ].join("\n");
    assert.deepStrictEqual(
      order(src, [2, 0, 1], ["lazyInitializer"]),
      [2, 0, 1]
    );
  });

  it("always returns a permutation of its input, whatever the desired order", () => {
    const src = [
      "var mod = {};",
      "function f() {}",
      "defineModuleExports(mod, {});",
      'var lit = "x";',
      "var derived = lit + mod;",
      "function g() {}"
    ].join("\n");
    const f = facts(src);
    const slots = f.map((_, i) => i);
    const next = seededShuffler(12345);
    for (let trial = 0; trial < 50; trial++) {
      const desired = next(slots);
      const out = orderRespectingLoadOrder(slots, desired, f);
      assert.deepStrictEqual(
        [...out].sort((a, b) => a - b),
        slots,
        `not a permutation for desired ${desired.join(",")}`
      );
      assertNoBarrierCrossed(out, f);
      assertNoReaderAboveWriter(out, f);
    }
  });

  it("returns bundle order unchanged when that is what is desired", () => {
    const src = ["var a = 1;", "boot();", "var b = 2;"].join("\n");
    assert.deepStrictEqual(order(src, [0, 1, 2]), [0, 1, 2]);
  });
});

/**
 * The bundler's export registrar, exactly as the bundle defines it. Detection is
 * STRUCTURAL, so the identifier names here are arbitrary on purpose — the real
 * ones are LLM-chosen and differ every run.
 */
const REGISTRAR = `var defineModuleExports = (targetObject, sourceObject) => {
    for (var propKey in sourceObject) defineProperty(targetObject, propKey, {
      get: sourceObject[propKey],
      enumerable: true,
      configurable: true,
      set: BoundIdentityProperty.bind(sourceObject, propKey)
    });
  };`;

describe("export-registrar calls — the largest self-pinned block of reorder churn", () => {
  it("is NOT an effect barrier, and WRITES its target rather than reading it", () => {
    // Its body installs LAZY getters (`get: source[k]`) over a literal of arrow
    // thunks, so nothing it is handed is evaluated at registration. Modelling it
    // as `pure` would be wrong in the other direction: `pure` records the target
    // as a READ, and two reads carry no edge, so a load-time read of
    // `exportsObj.foo` could then be ordered BEFORE the registration. A WRITE is
    // the honest model and gives read-after-write for free.
    const src = `${REGISTRAR}
      var exportsObj = {};
      defineModuleExports(exportsObj, { foo: () => 1 });`;
    const f = bundleLoadOrderFacts(stmts(src), src);
    const reg = f[2];
    assert.strictEqual(reg.effects, false, "not a barrier");
    assert.ok(reg.writes.includes("exportsObj"), "writes its target");
    assert.ok(!reg.reads.includes("exportsObj"), "does not merely read it");
  });

  it("still cannot float above its target's declaration (the boot-crash rule)", () => {
    // exp037 moved such a call above `var m = {}` and crashed the runnable tree.
    // The write-after-write edge is what forbids it, so ask the scheduler.
    const src = `${REGISTRAR}
      var exportsObj = {};
      defineModuleExports(exportsObj, { foo: () => 1 });`;
    const body = stmts(src);
    const f = bundleLoadOrderFacts(body, src);
    const slots = [0, 1, 2];
    // Desired order asks for the registration FIRST — illegal, must be refused.
    const got = orderRespectingLoadOrder(slots, [2, 1, 0], f);
    assert.ok(
      got.indexOf(1) < got.indexOf(2),
      `declaration must precede registration, got ${got.join(",")}`
    );
  });

  it("keeps a load-time reader of the target after the registration", () => {
    const src = `${REGISTRAR}
      var exportsObj = {};
      defineModuleExports(exportsObj, { foo: () => 1 });
      var copy = exportsObj.foo;`;
    const body = stmts(src);
    const f = bundleLoadOrderFacts(body, src);
    const got = orderRespectingLoadOrder([0, 1, 2, 3], [3, 2, 1, 0], f);
    assert.ok(
      got.indexOf(2) < got.indexOf(3),
      `registration must precede the read, got ${got.join(",")}`
    );
  });

  it("leaves an unrelated call a barrier — the exemption is not blanket", () => {
    const src = `${REGISTRAR}
      var exportsObj = {};
      doSomethingElse(exportsObj, { foo: () => 1 });`;
    const src2 = src;
    const f = bundleLoadOrderFacts(stmts(src2), src2);
    assert.strictEqual(
      f[2].effects,
      true,
      "an unverified call stays a barrier"
    );
  });
});

File: crates/humanify-core/src/emit/load_order/load_order_test.rs
//! Ported from src/split/load-order.test.ts, plus the Babel-shape cases
//! the oxc walk has to translate (parens, chains, meta properties, labels,
//! private names).

use std::collections::HashSet;

use oxc_allocator::Allocator;

use super::{
    LoadOrderFacts, LoadOrderOptions, analyze_load_order, bundle_load_order_facts,
    order_respecting_load_order,
};
use crate::ingest::Ingest;

fn facts_with(src: &str, pure: &[&str]) -> Vec<LoadOrderFacts> {
    let allocator = Allocator::default();
    let ingest = Ingest::parse(&allocator, src, "t.js");
    assert!(ingest.errors.is_empty(), "{:?}", ingest.errors);
    let opts = LoadOrderOptions {
        pure_call_names: pure.iter().map(|s| s.to_string()).collect(),
        target_writing_call_names: HashSet::new(),
    };
    analyze_load_order(&ingest.program.body, &opts)
}

fn facts(src: &str) -> Vec<LoadOrderFacts> {
    facts_with(src, &[])
}

fn bundle_facts(src: &str) -> Vec<LoadOrderFacts> {
    let allocator = Allocator::default();
    let ingest = Ingest::parse(&allocator, src, "t.js");
    assert!(ingest.errors.is_empty(), "{:?}", ingest.errors);
    bundle_load_order_facts(&ingest.program.body, src, false)
}

fn v(xs: &[&str]) -> Vec<String> {
    xs.iter().map(|s| s.to_string()).collect()
}

#[test]
fn a_function_declaration_is_hoisted_with_no_reads_writes_or_effects() {
    let f = &facts("function render(x) { return helper(x) + sideEffect(); }")[0];
    assert!(f.hoisted);
    assert!(!f.effects);
    assert!(f.reads.is_empty() && f.writes.is_empty());
}

#[test]
fn a_literal_declaration_is_a_pure_write() {
    let f = &facts("var TAG = \"[enforce]\";")[0];
    assert!(!f.hoisted && !f.effects);
    assert_eq!(f.writes, v(&["TAG"]));
    assert!(f.reads.is_empty());
}

#[test]
fn a_function_body_is_not_read_but_an_initializer_is() {
    let f = &facts("var run = () => { return moduleState + other(); };")[0];
    assert!(!f.effects);
    assert_eq!(f.writes, v(&["run"]));
    assert!(f.reads.is_empty());
    let g = &facts("var derived = base.field;")[0];
    assert_eq!(g.writes, v(&["derived"]));
    assert_eq!(g.reads, v(&["base"]));
}

#[test]
fn an_unknown_call_is_an_effect_and_a_verified_pure_one_is_not() {
    assert!(facts("var conn = openSocket();")[0].effects);
    let f = &facts_with(
        "var cfg = lazyInitializer(() => buildConfig(globalSettings));",
        &["lazyInitializer"],
    )[0];
    assert!(!f.effects);
    assert_eq!(f.writes, v(&["cfg"]));
    assert_eq!(f.reads, v(&["lazyInitializer"]));
}

#[test]
fn the_pure_wrapper_is_recognised_through_the_sequence_callee_form() {
    let f = &facts_with(
        "var cfg = (0, resourceLifecycle.lazyInitializer)(() => build());",
        &["lazyInitializer"],
    )[0];
    assert!(!f.effects);
    assert_eq!(f.reads, v(&["resourceLifecycle"]));
    let g = &facts_with(
        "var cfg = (0, resourceLifecycle.runNow)(() => build());",
        &["lazyInitializer"],
    )[0];
    assert!(g.effects);
}

#[test]
fn bare_var_is_free_and_bare_let_is_a_write() {
    let f = &facts("var pending, queued;")[0];
    assert!(!f.effects);
    assert!(f.writes.is_empty());
    assert_eq!(facts("let pending;")[0].writes, v(&["pending"]));
}

#[test]
fn registration_calls_assignments_and_spreads() {
    let f = &facts("defineModuleExports(memoryModule, { get: () => x });")[0];
    assert!(f.effects);
    assert!(f.reads.contains(&"memoryModule".to_string()));
    let g = &facts("cachedValue = 1;")[0];
    assert!(!g.effects);
    assert_eq!(g.writes, v(&["cachedValue"]));
    assert!(facts("target.field = 1;")[0].effects);
    assert!(facts("var merged = { ...source };")[0].effects);
}

#[test]
fn classes_and_control_flow() {
    let f = &facts("class Panel extends Base { render() { return sideFx(); } }")[0];
    assert!(!f.effects);
    assert_eq!(f.writes, v(&["Panel"]));
    assert_eq!(f.reads, v(&["Base"]));
    assert!(facts("class Registry { static all = collectAll(); }")[0].effects);
    assert!(facts("if (flag) { boot(); }")[0].effects);
    assert!(facts("try { boot(); } catch {}")[0].effects);
    assert!(facts("for (var i = 0; i < 3; i++) tick();")[0].effects);
}

#[test]
fn babel_identifiers_the_oxc_shape_hides_are_reads() {
    // A label, a catch param and a nested declarator id are all Babel
    // Identifier nodes the generic walk reaches.
    let f = &facts("outer: for (var k of list) { break outer; }")[0];
    assert_eq!(f.reads, v(&["outer", "k", "list"]));
    let g = &facts("try { a(); } catch (err) { b(err); }")[0];
    assert_eq!(g.reads, v(&["a", "err", "b"]));
}

#[test]
fn parens_and_optional_chains_are_transparent() {
    let f = &facts_with("var x = (lazy)(() => 1);", &["lazy"])[0];
    assert!(!f.effects);
    let g = &facts_with("var y = ns?.lazy(() => 1);", &["lazy"])[0];
    assert!(!g.effects);
    assert_eq!(g.reads, v(&["ns"]));
}

#[test]
fn destructuring_targets_read_and_destructuring_declarations_are_effects() {
    let f = &facts("({ a, b: c } = src);")[0];
    assert!(f.effects);
    assert_eq!(f.reads, v(&["a", "c", "src"]));
    let g = &facts("var { p, q } = obj;")[0];
    assert!(g.effects);
    assert_eq!(g.writes, v(&["p", "q"]));
}

const HELPER: &str = "var qz = (gen, cached) => () => (gen && (cached = gen(gen = 0)), cached);";

#[test]
fn the_lazy_init_helper_is_verified_by_shape_not_by_name() {
    let src = format!("{HELPER}\nvar mod = qz(() => loadHeavyThing());");
    let f = bundle_facts(&src);
    assert!(!f[1].effects);
    assert_eq!(f[1].writes, v(&["mod"]));
    let src2 = "var qz = makeThing;\nvar mod = qz(() => loadHeavyThing());";
    assert!(bundle_facts(src2)[1].effects);
}

const REGISTRAR: &str = "var defineModuleExports = (targetObject, sourceObject) => {
    for (var propKey in sourceObject) defineProperty(targetObject, propKey, {
      get: sourceObject[propKey],
      enumerable: true,
      configurable: true,
      set: BoundIdentityProperty.bind(sourceObject, propKey)
    });
  };";

#[test]
fn the_export_registrar_writes_its_target_and_is_not_a_barrier() {
    let src = format!(
        "{REGISTRAR}\nvar exportsObj = {{}};\ndefineModuleExports(exportsObj, {{ foo: () => 1 }});"
    );
    let f = bundle_facts(&src);
    assert!(!f[2].effects);
    assert!(f[2].writes.contains(&"exportsObj".to_string()));
    assert!(!f[2].reads.contains(&"exportsObj".to_string()));
    let got = order_respecting_load_order(&[0, 1, 2], &[2, 1, 0], &f);
    let pos = |x: usize| got.iter().position(|&s| s == x).unwrap();
    assert!(
        pos(1) < pos(2),
        "declaration must precede registration: {got:?}"
    );
}

#[test]
fn a_load_time_reader_of_the_registrar_target_stays_after_it() {
    let src = format!(
        "{REGISTRAR}\nvar exportsObj = {{}};\ndefineModuleExports(exportsObj, {{ foo: () => 1 }});\nvar copy = exportsObj.foo;"
    );
    let f = bundle_facts(&src);
    let got = order_respecting_load_order(&[0, 1, 2, 3], &[3, 2, 1, 0], &f);
    let pos = |x: usize| got.iter().position(|&s| s == x).unwrap();
    assert!(pos(2) < pos(3), "{got:?}");
}

#[test]
fn an_unrelated_call_stays_a_barrier() {
    let src = format!(
        "{REGISTRAR}\nvar exportsObj = {{}};\ndoSomethingElse(exportsObj, {{ foo: () => 1 }});"
    );
    assert!(bundle_facts(&src)[2].effects);
}

fn order(src: &str, desired: &[usize], pure: &[&str]) -> Vec<usize> {
    let f = facts_with(src, pure);
    let slots: Vec<usize> = (0..f.len()).collect();
    order_respecting_load_order(&slots, desired, &f)
}

#[test]
fn pure_declarations_permute_freely() {
    let src = "var a = \"one\";\nvar b = 2;\nvar c = () => {};\nvar d = /re/;";
    for desired in [[3, 2, 1, 0], [1, 3, 0, 2], [2, 0, 3, 1]] {
        assert_eq!(order(src, &desired, &[]), desired.to_vec());
    }
}

#[test]
fn dependences_and_barriers_hold() {
    assert_eq!(
        order("var total = 5;\nvar doubled = total * 2;", &[1, 0], &[]),
        vec![0, 1]
    );
    assert_eq!(
        order("var view = state;\nvar state = compute;", &[1, 0], &[]),
        vec![0, 1]
    );
    assert_eq!(
        order("counter = 1;\ncounter = 2;", &[1, 0], &[]),
        vec![0, 1]
    );
    assert_eq!(order("boot();\nshutdown();", &[1, 0], &[]), vec![0, 1]);
    let crash = "var mod = {};\ndefineModuleExports(mod, { get: () => v });\nvar tail = \"t\";";
    assert_eq!(order(crash, &[2, 1, 0], &[]), vec![0, 1, 2]);
    assert_eq!(order(crash, &[1, 0, 2], &[]), vec![0, 1, 2]);
}

#[test]
fn a_hoisted_function_crosses_a_barrier() {
    let src = "var mod = {};\ndefineModuleExports(mod, {});\nfunction tail() {}";
    assert_eq!(order(src, &[2, 0, 1], &[]), vec![2, 0, 1]);
}

#[test]
fn pure_lazy_init_declarations_move() {
    let src = "var first = lazyInitializer(() => a());\nvar second = lazyInitializer(() => b());\nvar third = lazyInitializer(() => c());";
    assert_eq!(order(src, &[2, 0, 1], &["lazyInitializer"]), vec![2, 0, 1]);
}

#[test]
fn always_a_permutation_that_crosses_no_barrier() {
    let src = "var mod = {};\nfunction f() {}\ndefineModuleExports(mod, {});\nvar lit = \"x\";\nvar derived = lit + mod;\nfunction g() {}";
    let f = facts(src);
    let slots: Vec<usize> = (0..f.len()).collect();
    // The TS's seeded shuffler (an LCG), 50 trials.
    let mut state: u64 = 12345;
    for _ in 0..50 {
        let mut desired = slots.clone();
        for i in (1..desired.len()).rev() {
            state = (state * 1_103_515_245 + 12_345) % 2_147_483_648;
            let j = ((state as f64 / 2_147_483_648.0) * (i + 1) as f64).floor() as usize;
            desired.swap(i, j);
        }
        let out = order_respecting_load_order(&slots, &desired, &f);
        let mut sorted = out.clone();
        sorted.sort_unstable();
        assert_eq!(sorted, slots);
        let pos = |x: usize| out.iter().position(|&s| s == x).unwrap();
        for &b in &out {
            if !f[b].effects {
                continue;
            }
            for &s in &out {
                if s != b && !f[s].hoisted {
                    assert_eq!(s < b, pos(s) < pos(b), "{s} crossed barrier {b}");
                }
            }
        }
    }
}
